docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+Because kubeadm is platform-independent and is intended to provide an “easy path” to cluster bootstrapping,  there are a number of inherent design decisions that limit the overall security of the provisioned cluster:
+- Kubeadm uses TLS bootstrapping for node registration, however the default workflow used by Cluster API uses bootstrap token which allow registration as arbitrary node names.
+    - When used in this mode, Kubeadm essentially does “client-side validation” to prevent node hijacking, but this does not mean the token cannot be reused by an attacker within the lifetime of the token to perform a hijack.
+    - Cluster API cannot scope a token down  to a specific node, because bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
+- Kubeadm sets node labels during bootstrapping, and Cluster API does not do reconciliation of node labels post node deployment.
+    - This is helpful because it means the kubelet starts with the labels applied straightaway, preventing a race condition when the Kubernetes schedule may misplace a pod due to a temporarily missing node label.
+    - However, again, this allows for potential hijacking, for example, using the kubelet’s credentials to apply a label intended for workloads not using the kubernetes.io/node-restriction-* labels.
+
+
+### Goals
+
+- Provide a bootstrap mechanism that assures node secure node registration
+- The node registration 
+
+
+### Non-Goals/Future Work
+
+- Not to change assumptions around management cluster to workload cluster connectivity
+- Not to solve the protection of initial cluster bootstrap secrets for the control plane nodes (see Node Bootstrapper)
+
+## Proposal
+
+
+
+| Component | Location | Description |
+| --------- | -------- | ----------- |
+| node-attestation-controller     | Part of Cluster API repository, under bootstrap/node/attestation, and imported by Cluster API infra providers for implementation. A generic challenge-response implementation will be included for providers / bare metal without an attestation mechanism.     |   A controller to verify and sign the CSR. this would be typically an importable controller where the infrastructure provider implements the interface with specific code for CSR approval + Start the controller as part of its main.go or through an independent binary.   |
+
+#### Kubelet authentication plugin
+
+![sequence diagram](images/kubelet-authentication/bootstrap-flow.png)
+
+
+
+The following sequence diagram shows at a high level how a nodeConfig enables bootstrapping a kubernetes node. This sequence diagram is independent of API definition and alternatives and is relevant in all cases.
+
+We propose a kubelet authentication plugin to be present on the instances, (how the nodeadm CLI made available is out of scope for this proposal), which will be responsible for node registration, as well as certificate rotation. The agent will be made up of two parts:
```

Comment:
> how the nodeadm CLI made available is out of scope for this proposal

this needs to be defined in future goals if it's out of scope

Reply:
I need to reword this, make it downloadable on the fly is out of scope, how it's made available initially is going to be through image-builder